### PR TITLE
Remove DummyCreationSession

### PR DESCRIPTION
**Pull request by blairconrad**
Connects to #1489.

I noticed that we check for recursive fake creation even before looking in the cache for creation strategies. If there's a strategy in the cache, we can't have a recursive call.
So I rearranged those calls and then noted that most of the time we don't need a DummyCreationSession.
I looked at ways to not create one until we needed it, and nothing felt right until I just folded it into the DummyValueResolver. Then none of the clients needed to know about the session at all.

I left the changes unsquashed in https://github.com/blairconrad/FakeItEasy/tree/no-session-unsquashed in case you find this diff hard to follow.

|          Method |               Job |      Mean |     Error |    StdDev | Ratio | RatioSD | Gen 0/1k Op | Gen 1/1k Op | Gen 2/1k Op | Allocated Memory/Op |
|---------------- |------------------ |----------:|----------:|----------:|------:|--------:|------------:|------------:|------------:|--------------------:|
|        Callback |   4.10.0-alpha091 | 21.189 us | 1.0675 us | 0.9985 us |  1.00 |    0.00 |      1.8600 |      0.0200 |      0.0100 |              5958 B |
|        Callback | 4.10.0-no-session | 21.281 us | 0.6127 us | 0.5731 us |  1.01 |    0.05 |      1.7600 |      0.0200 |      0.0100 |              5668 B |
|    CallbackOnly |   4.10.0-alpha091 |  2.815 us | 0.0688 us | 0.0644 us |  1.00 |    0.00 |      0.1700 |      0.0600 |           - |              1096 B |
|    CallbackOnly | 4.10.0-no-session |  2.654 us | 0.0711 us | 0.0665 us |  0.94 |    0.03 |      0.1400 |      0.0700 |           - |               896 B |
|    Construction |   4.10.0-alpha091 |  8.348 us | 1.3449 us | 1.2580 us |  1.00 |    0.00 |      1.0300 |      0.0100 |           - |              3361 B |
|    Construction | 4.10.0-no-session |  7.677 us | 0.2316 us | 0.2167 us |  0.93 |    0.11 |      1.0100 |      0.0100 |           - |              3273 B |
|     EmptyMethod |   4.10.0-alpha091 |  9.801 us | 0.1896 us | 0.1774 us |  1.00 |    0.00 |      1.3000 |      0.0100 |           - |              4187 B |
|     EmptyMethod | 4.10.0-no-session |  9.618 us | 0.0907 us | 0.0849 us |  0.98 |    0.02 |      1.2100 |      0.0100 |           - |              3897 B |
| EmptyMethodOnly |   4.10.0-alpha091 |  2.607 us | 0.0623 us | 0.0582 us |  1.00 |    0.00 |      0.1500 |      0.0500 |           - |               952 B |
| EmptyMethodOnly | 4.10.0-no-session |  2.349 us | 0.0443 us | 0.0414 us |  0.90 |    0.03 |      0.1200 |      0.0400 |           - |               752 B |
|    EmptyReturns |   4.10.0-alpha091 | 10.570 us | 1.2906 us | 1.2073 us |  1.00 |    0.00 |      1.3300 |      0.0100 |           - |              4275 B |
|    EmptyReturns | 4.10.0-no-session |  9.887 us | 0.2267 us | 0.2120 us |  0.94 |    0.08 |      1.2400 |      0.0100 |           - |              3986 B |
| EmptyReturnOnly |   4.10.0-alpha091 |  3.092 us | 0.0915 us | 0.0856 us |  1.00 |    0.00 |      0.1700 |      0.0500 |           - |              1040 B |
| EmptyReturnOnly | 4.10.0-no-session |  2.730 us | 0.1569 us | 0.1467 us |  0.88 |    0.04 |      0.1300 |      0.0500 |           - |               840 B |
|          Return |   4.10.0-alpha091 | 22.019 us | 2.0976 us | 1.9621 us |  1.00 |    0.00 |      1.7900 |      0.0200 |      0.0100 |              5773 B |
|          Return | 4.10.0-no-session | 20.331 us | 0.1455 us | 0.1361 us |  0.93 |    0.08 |      1.7700 |      0.0200 |      0.0100 |              5684 B |
|      ReturnOnly |   4.10.0-alpha091 |  3.640 us | 0.5267 us | 0.4927 us |  1.00 |    0.00 |      0.1700 |      0.0500 |           - |              1040 B |
|      ReturnOnly | 4.10.0-no-session |  2.689 us | 0.1516 us | 0.1418 us |  0.75 |    0.08 |      0.1300 |      0.0500 |           - |               840 B |

image

### CODE REVIEW

File: src/FakeItEasy/Creation/DummyValueResolver.cs [hunk 1]
```diff
@@ -26,57 +26,86 @@ public DummyValueResolver(DynamicDummyFactory dummyFactory, IFakeObjectCreator f
...
+                return cachedStrategy.TryCreateDummyValue(typeOfDummy, this);
             }
 
-            return creationResult;
+            var resolver = new LoopDetectingResovler(this.strategies, this.strategyCache);
+            return resolver.TryResolveDummyValueWithAllAvailableStrategies(typeOfDummy);
         }
 
-        private CreationResult TryResolveDummyValueWithAllAvailableStrategies(DummyCreationSession session, Type typeOfDummy)
+        private class LoopDetectingResovler : IDummyValueResolver
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DummyValueResolver.cs]: Typo: Resovler => Resolver

File: src/FakeItEasy/Creation/DummyValueResolver.cs [hunk 2]
```diff
@@ -26,57 +26,86 @@ public DummyValueResolver(DynamicDummyFactory dummyFactory, IFakeObjectCreator f
...
+            private readonly ResolveStrategy[] strategies;
+            private readonly ConcurrentDictionary<Type, ResolveStrategy> strategyCache;
+            private readonly HashSet<Type> typesCurrentlyBeingResolved;
+
+            public LoopDetectingResovler(ResolveStrategy[] strategies, ConcurrentDictionary<Type, ResolveStrategy> strategyCache)
             {
-                return cachedStrategy.TryCreateDummyValue(session, typeOfDummy);
+                this.strategies = strategies;
+                this.strategyCache = strategyCache;
+                this.typesCurrentlyBeingResolved = new HashSet<Type>();
```
Comment (thomaslevesque) [inline src/FakeItEasy/Creation/DummyValueResolver.cs]: Isn't creating this immediately basically the same as creating the `DummyCreationSession` eagerly?
Maybe it could be created lazily; not sure how much of a benefit that would be.

File: src/FakeItEasy/Creation/DummyValueResolver.cs (hunk 1 above)
Comment (blairconrad) [reply src/FakeItEasy/Creation/DummyValueResolver.cs]: Thank you.

File: src/FakeItEasy/Creation/DummyValueResolver.cs (hunk 2 above)
Comment (blairconrad) [reply src/FakeItEasy/Creation/DummyValueResolver.cs]: Ah, no. In many many cases (I hope), we'll be creating a Dummy of a type we've attempted to use before, so the `DummyValueResolver` will find a strategy in its cache and create the Dummy without ever making a `LoopDetectingResolver`.
However, when there's no cached strategy, we create `LoopDetectingResolver` and the `DummyValueResolver` calls `TryResolveDummyValueWithAllAvailableStrategies` immediately, so `typesCurrentlyBeingResolved` will be used.

**Comment by blairconrad** [comment]
I fixed the typo and repushed (straight amend, no fixup). I hope my eager/lazy comment made sense.

File: src/FakeItEasy/Creation/DummyValueResolver.cs (hunk 2 above)
Comment (thomaslevesque) [reply src/FakeItEasy/Creation/DummyValueResolver.cs]: Ah, I see

**Comment by thomaslevesque** [comment]
Thanks @blairconrad, nice work!

**Comment by blairconrad** [comment]
Thank you. Glad you caught that typo!
